esp-hal/src/work_queue.rs

```diff
@@ -0,0 +1,459 @@
+/// These functions may be called both in the context of the queue frontends and drivers.
+pub(crate) struct VTable<T: Sync> {
+    /// Starts processing a new work item.
+    ///
+    /// The function returns whether the work item was accepted. If there is no driver currently
+    /// processing the queue, this function will return false to prevent removing the work item
+    /// from the queue.
+    ///
+    /// This function should be as short as possible.
+    pub(crate) post: fn(*const (), &mut T) -> bool,
+
+    /// Polls the status of a particular work item.
+    ///
+    /// Return `None` if the work item is not currently being processed.
+    ///
+    /// This function should be as short as possible.
+    pub(crate) poll: fn(*const (), &mut T) -> Option<Poll>,
+
+    /// Attempts to abort processing a work item.
+    ///
+    /// This function should be as short as possible.
+    pub(crate) cancel: fn(*const (), &mut T),
+
+    /// Called when the driver may be stopped.
+    ///
+    /// This function should be as short as possible.
+    pub(crate) stop: fn(*const ()),
+}
+
+impl<T: Sync> VTable<T> {
+    pub(crate) const fn noop() -> Self {
+        Self {
+            post: |_, _| false,
+            poll: |_, _| None,
+            cancel: |_, _| (),
+            stop: |_| (),
+        }
+    }
+}
+
+struct Inner<T: Sync> {
+    head: Option<NonNull<WorkItem<T>>>,
+    tail: Option<NonNull<WorkItem<T>>>,
+    current: Option<NonNull<WorkItem<T>>>,
+
+    // The data pointer will be passed to VTable functions, which may be called in any context.
+    data: *const (),
+    vtable: VTable<T>,
+}
+
+impl<T: Sync> Inner<T> {
+    /// Places a work item at the end of the queue.
+    fn enqueue(&mut self, ptr: NonNull<WorkItem<T>>) {
+        if let Some(tail) = self.tail.as_mut() {
+            // Queue contains something, append to `tail`.
+            unsafe { tail.as_mut().next = Some(ptr) };
+        } else {
+            // Queue was empty, set `head` to the first element.
+            self.head = Some(ptr);
+        }
+
+        // Move `tail` to the newly inserted item.
+        self.tail = Some(ptr);
+    }
+
+    /// Places a work item at the front of the queue.
+    fn enqueue_front(&mut self, mut ptr: NonNull<WorkItem<T>>) {
+        // Chain the node into the list.
+        unsafe { ptr.as_mut().next = self.head };
+
+        // Adjust list `head` to point at the new-first element.
+        self.head = Some(ptr);
+        if self.tail.is_none() {
+            // The queue was empty, we need to set `tail` to the last element.
+            self.tail = Some(ptr);
+        }
+    }
+
+    /// Runs one processing iteration.
+    ///
+    /// This function enqueues a new work item or polls the status of the currently processed one.
+    fn process(&mut self) {
+        if let Some(mut current) = self.current {
+            let result = unsafe { (self.vtable.poll)(self.data, &mut current.as_mut().data) };
```

Comment:
This is all you need right?
```suggestion
            let result = (self.vtable.poll)(self.data, unsafe { &mut current.as_mut().data });
```